### DISCUSSION TITLE
Does PascalABC.NET support Delphi variant records?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3071] [Q&A] [answered]
https://github.com/vtereshkov/xdp/blob/master/xdp.dpr#L216-L229

**Comment by SunSerega** [comment, marked as answer]
Generally, you should use OOP things for this, like `abstract class`es and `interface`s.
But when you are memory- or performance-critical, use:
```
  [StructLayout(LayoutKind.Explicit)]
  my_union = record
    [FieldOffset(0)] data1: real;
    [FieldOffset(0)] data2: int64;
    [FieldOffset(0)] data3: record
      i1, i2: integer;
    end;
  end;
```

**Reply by ghost** [reply]
Is this the same as http://docwiki.embarcadero.com/RADStudio/Alexandria/en/Structured_Types_(Delphi)#Variant_Parts_in_Records?

**Reply by SunSerega** [reply]
No, it's closer to unions from C/C++.
Basically, I declared 3 fields that are all placed on top of each other, in the same memory location.
But you can simulate Delphi variants using unions: by making another `record` with a field of this union type + a field to choose which field in the union currently holds the valid info.

**Reply by ghost** [reply]
> But you can simulate Delphi variants using unions: by making another `record` with a field of this union type + a field to choose which field in the union currently holds the valid info.

@SunSerega Could you give me code? I can't imagine how it will look like. Thank you.

**Reply by SunSerega** [reply]
I heard you the first time, without the ping - no need to recreate your message. Well, I'm still to busy to make detailed answer.
But I guess you can't wait - so here's an idea I said but in code, so you can at least imagine what to do:
```
type
  [StructLayout(LayoutKind.Explicit)]
  my_union = record
    [FieldOffset(0)] data1: real;
    [FieldOffset(0)] data2: int64;
    [FieldOffset(0)] data3: record
      i1, i2: integer;
    end;
  end;
  
  my_variant_record = record
    u: my_union;
    v: byte;
    
    constructor(data1: real);
    begin
      u.data1 := data1;
      v := 1; // id of current main representation
    end;
    
    //TODO The rest of stuff
    // - Maybe constructor and property accepting "object" and using "match"
    
  end;
```
Please try to figure out at least something from here. If you are stuck - at least describe what you found and tried.
Making detailed descriptions of that in words tends to solve many problems (the rubber duck method)
And when it doesn't solve the problem - it at least makes it a much better question.
Making good answer to "I can't do anything" type of question is hard and demoralizing...
